distributed/semaphore.py

```diff
@@ -0,0 +1,336 @@
+                )
+
+    def _release_value(self, name, client, identifier):
+        # Everything needs to be atomic here.
+        self.leases_per_client[client][name].remove(identifier)
+        self.leases[name].remove(identifier)
+        self.events[name].set()
+
+    def _release_client(self, client):
+        semaphore_names = list(self.leases_per_client[client])
+        for name in semaphore_names:
+            ids = list(self.leases_per_client[client][name])
+            for _id in list(ids):
+                self._release_value(name=name, client=client, identifier=_id)
+        del self.leases_per_client[client]
+
+    def _validate_leases(self):
+        if not self._validation_running:
+            self._validation_running = True
+            known_clients_with_leases = set(self.leases_per_client.keys())
+            scheduler_clients = set(self.scheduler.clients.keys())
+            for dead_client in known_clients_with_leases - scheduler_clients:
+                self._release_client(dead_client)
+            else:
+                self._validation_running = False
+
+    def close(self, comm=None, name=None):
+        """Hard close the semaphore without warning clients which still hold a lease."""
+        with log_errors():
+            if not self._semaphore_exists(name):
+                return
+
+            del self.max_leases[name]
+            if name in self.events:
+                del self.events[name]
+            if name in self.leases:
+                del self.leases[name]
+
+            for client, client_leases in self.leases_per_client.items():
+                if name in client_leases:
+                    warnings.warn(
+                        f"Closing semaphore `{name}` but client `{client}` still has a lease open.",
+                        RuntimeWarning,
+                    )
+                    del client_leases[name]
+
+
+class Semaphore:
+    """ Semaphore
+
+    Parameters
+    ----------
+    max_leases: int (optional)
+        The maximum amount of leases that may be granted at the same time. This
+        effectively sets an upper limit to the amount of parallel access to a specific resource.
+        Defaults to 1.
+    name: string (optional)
+        Name of the semaphore to acquire.  Choosing the same name allows two
+        disconnected processes to coordinate.  If not given, a random
+        name will be generated.
+    client: Client (optional)
+        Client to use for communication with the scheduler.  If not given, the
+        default global client will be used.
+
+    Examples
+    --------
+    >>> from distributed import Semaphore
+    >>> sem = Semaphore(max_leases=2, name='my_database')
+    >>> # This automatically acquires a lease from the semaphore (if available) which will be released when leaving
+    >>> # the context manager
+    >>> with sem:
+    >>>     pass  # use the protected resource
+    >>> sem.close()  # Close resource (clearing any residual state)
```

Comment:
```suggestion
    >>> def access_ressource(s, sem):
    >>>     # This automatically acquires a lease from the semaphore (if available) which will be
    >>>     # released when leaving the context manager.
    >>>     with sem:
    >>>         pass
    >>>    
    >>> futures = client.map(access_ressource, range(10), sem=sem)
    >>> client.gather(futures)
    >>> # Once done, close the semaphore to clean up the state on scheduler side.
    >>> sem.close()
    
```

This example shows the (recommended) usage where the semaphore is initialised on client side and not in the worker. This usage makes the cleanup more straight forward